### PR TITLE
[workingset] verify correct tx after unfolding

### PR DESCRIPTION
**Pull request by dustinxie**
…legacy tx

# Description
Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/sealedenvelope.go [hunk 1]
```diff
@@ -242,3 +242,18 @@ func (sealed *SealedEnvelope) VerifySignature() error {
...
+		return errors.Wrapf(ErrInvalidAct, "invalid tx type = %d not supported", version)
+	}
+	if sealed.encoding != iotextypes.Encoding_IOTEX_PROTOBUF && sealed.encoding != iotextypes.Encoding_ETHEREUM_EIP155 &&
+		sealed.encoding != iotextypes.Encoding_ETHEREUM_UNPROTECTED && sealed.encoding != iotextypes.Encoding_TX_CONTAINER {
+		return errors.Wrapf(ErrInvalidAct, "invalid tx encoding = %d not supported", sealed.encoding)
+	}
+	if sealed.encoding == iotextypes.Encoding_IOTEX_PROTOBUF && sealed.Version() != LegacyTxType {
+		return errors.Wrap(ErrInvalidAct, "protobuf encoding only supports legacy tx")
+	}
+	return nil
```
Comment (CoderZhi) [inline action/sealedenvelope.go:258]: This is more like a logic which should be in GenericValidator

File: action/sealedenvelope.go [hunk 2]
```diff
@@ -242,3 +242,18 @@ func (sealed *SealedEnvelope) VerifySignature() error {
 	}
 	return nil
 }
+
+func (sealed *SealedEnvelope) ValidateVersionEncoding() error {
+	if version := sealed.Version(); version != LegacyTxType && version != AccessListTxType &&
```
Comment (CoderZhi) [inline action/sealedenvelope.go:247]: https://github.com/iotexproject/iotex-core/pull/4432#discussion_r1800477909

File: action/sealedenvelope.go [hunk 3]
```diff
@@ -242,3 +242,18 @@ func (sealed *SealedEnvelope) VerifySignature() error {
...
+		version != DynamicFeeTxType && version != BlobTxType {
+		return errors.Wrapf(ErrInvalidAct, "invalid tx type = %d not supported", version)
+	}
+	if sealed.encoding != iotextypes.Encoding_IOTEX_PROTOBUF && sealed.encoding != iotextypes.Encoding_ETHEREUM_EIP155 &&
```
Comment (CoderZhi) [inline action/sealedenvelope.go:251]: `switch` may be easier to read

File: action/sealedenvelope.go (hunk 1 above)
Comment (envestcc) [reply action/sealedenvelope.go:258]: agree

**Review by envestcc — APPROVED** [review]

File: state/factory/workingset.go [hunk 4]
```diff
@@ -663,12 +675,19 @@ func (ws *workingSet) pickAndRunActions(
 			if fCtx.UnfoldContainerBeforeValidate {
 				if container, ok := nextAction.Envelope.(action.TxContainer); ok {
 					if err := container.Unfold(nextAction, ctx, ws.checkContract); err != nil {
-						if caller := nextAction.SenderAddress(); caller != nil {
-							ap.DeleteAction(caller)
-						}
-						actionIterator.PopAccount()
+						removeCaller(ap, actionIterator, nextAction.SenderAddress())
```
Comment (CoderZhi) [inline state/factory/workingset.go:678]: print the error via log.Debug

File: state/factory/workingset.go [hunk 5]
```diff
@@ -663,12 +675,19 @@ func (ws *workingSet) pickAndRunActions(
...
 						continue
 					}
+					if nextAction.To() != nil {
+						var (
+							ioAddr, _ = address.FromBytes(nextAction.To().Bytes())
+							to        = ioAddr.String()
+						)
+						if to == address.StakingProtocolAddr || to == address.RewardingProtocol {
```
Comment (CoderZhi) [inline state/factory/workingset.go:686]: how is this possible in pickAndRun?

File: action/protocol/generic_validator.go [hunk 6]
```diff
@@ -94,6 +95,10 @@ func (v *GenericValidator) Validate(ctx context.Context, selp *action.SealedEnve
 		if ok && !featureCtx.EnableBlobTransaction && selp.TxType() == action.BlobTxType {
 			return errors.Wrap(action.ErrInvalidAct, "blob tx is not enabled")
 		}
+		if ok && featureCtx.EnableNewTxTypes && selp.TxType() != action.LegacyTxType &&
+			selp.Encoding() == uint32(iotextypes.Encoding_IOTEX_PROTOBUF) {
+			return errors.Wrap(action.ErrInvalidAct, "protobuf encoding only supports legacy tx")
```
Comment (dustinxie) [inline action/protocol/generic_validator.go:100]: make sure protobuf only support `LegacyTx`

File: state/factory/workingset.go [hunk 7]
```diff
@@ -526,11 +530,13 @@ func (ws *workingSet) processWithCorrectOrder(ctx context.Context, actions []*ac
 			return errors.Wrap(err, "error when run action")
 		}
 		receipts = append(receipts, receipt)
-		blkCtx.GasLimit -= receipt.GasConsumed
-		if fCtx.EnableDynamicFeeTx && receipt.PriorityFee() != nil {
-			(&blkCtx.AccumulatedTips).Add(&blkCtx.AccumulatedTips, receipt.PriorityFee())
+		if !action.IsSystemAction(act) {
```
Comment (dustinxie) [inline state/factory/workingset.go:533]: after discussion, it does not hurt to add this check, to make sure the behavior is the same as`pickAndRunActions()`

File: action/sealedenvelope.go (hunk 1 above)
Comment (dustinxie) [reply action/sealedenvelope.go]: fixed in latest commit

File: state/factory/workingset.go (hunk 5 above)
Comment (dustinxie) [reply state/factory/workingset.go:686]: yeah, after discussion, this is not very meaningful, a malicious node can anyway modify his `pickAndRun` to send such a tx in container format, so removed

File: state/factory/workingset.go [hunk 8]
```diff
@@ -663,10 +675,8 @@ func (ws *workingSet) pickAndRunActions(
 			if fCtx.UnfoldContainerBeforeValidate {
 				if container, ok := nextAction.Envelope.(action.TxContainer); ok {
 					if err := container.Unfold(nextAction, ctx, ws.checkContract); err != nil {
-						if caller := nextAction.SenderAddress(); caller != nil {
-							ap.DeleteAction(caller)
-						}
-						actionIterator.PopAccount()
+						log.L().Info("failed to unfold tx container", zap.Uint64("height", ws.height), zap.Error(err))
```
Comment (CoderZhi) [inline state/factory/workingset.go:678]: change to Debug

File: state/factory/workingset.go [hunk 9]
```diff
@@ -681,13 +691,13 @@ func (ws *workingSet) pickAndRunActions(
...
 			}
+			caller := nextAction.SenderAddress()
 			if err != nil {
-				caller := nextAction.SenderAddress()
 				if caller == nil {
 					return nil, errors.New("failed to get address")
 				}
-				ap.DeleteAction(caller)
-				actionIterator.PopAccount()
+				log.L().Info("failed to validate tx", zap.Uint64("height", ws.height), zap.Error(err))
```
Comment (CoderZhi) [inline state/factory/workingset.go:699]: change to Debug

File: state/factory/workingset.go [hunk 10]
```diff
@@ -698,16 +708,11 @@ func (ws *workingSet) pickAndRunActions(
 				actionIterator.PopAccount()
 				continue
 			case action.ErrChainID, errUnfoldTxContainer, errDeployerNotWhitelisted:
-				if caller := nextAction.SenderAddress(); caller != nil {
-					ap.DeleteAction(caller)
-				}
-				actionIterator.PopAccount()
+				log.L().Info("runAction() failed", zap.Uint64("height", ws.height), zap.Error(err))
```
Comment (CoderZhi) [inline state/factory/workingset.go:711]: change to Debug

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**